### PR TITLE
VLC player in background

### PR DESCRIPTION
**Pull request by JDuqn**
Joined PRs :
- https://bitbucket.taktik.be/projects/FLOWR/repos/flowr/pull-requests/4663/overview
- https://github.com/taktik/flowr-desktop/pull/40
- https://bitbucket.taktik.be/projects/FLOWR/repos/bedside-vlc/pull-requests/1/overview

### CODE REVIEW

File: src/frontend/src/players/abstractPlayer.ts [hunk 1]
```diff
@@ -57,6 +57,7 @@ abstract class AbstractPlayer implements IPlayer {
     // eslint-disable-next-line @typescript-eslint/no-unsafe-return
     const playerConfigMerged = mergeWith({}, DEFAULT_PLAYER_STORE, playerConfig, (a, b) => b === null || b === '' ? a : undefined)
     this.store = storeManager.createStore<IPlayerStore>('player', playerConfigMerged)
+    this.store.persist()
```
Comment (lmancel) [inline src/frontend/src/players/abstractPlayer.ts]: Not necessary anymore, the store is persisted upon creation "automatically"

File: src/frontend/index.ts [hunk 2]
```diff
@@ -73,6 +73,8 @@ export async function createFlowrWindow(flowrStore: Store<IFlowrStore>): Promise
       partition: 'persist:flowr', // needed to display webcam image
       preload: buildPreloadPath('exportNode.js'),
     },
+    transparent: true,
+    frame: false
```
Comment (lmancel) [inline src/frontend/index.ts]: Both values should not be hardcoded: they must depend on the value of vlc player's position.
Especially the `frame: false` that would completely remove the possibility to drag flowr-desktop's window.
This implies that if we notice that we did not open the window with the settings that match the ones received from flowr-frontend, then flowr's window must be destroyed and created again

**Review by lmancel — CHANGES_REQUESTED** [review]

**Comment by Azevedo-Dani** [comment]
We store this settings in flowr-admin. We should also store that in user-preferences.json no ?

If no, there's a logic to know which setting we want store or not in user-preferences

**Comment by lmancel** [comment]
> We store this settings in flowr-admin. We should also store that in user-preferences.json no ?
> 
> If no, there's a logic to know which setting we want store or not in user-preferences

It's defined in the player's store, and thus in player.json. Nothing else to be done.

File: src/frontend/src/players/vlc/player.ts [hunk 3]
```diff
@@ -53,12 +53,12 @@ export class VlcPlayer extends AbstractPlayer {
   }
 
   private startProcess(url: string) {
-    const path = this.store.get('pipeline')?.metadata?.applicationPath
-
+    //const path = this.store.get('pipeline')?.metadata?.applicationPath
+    const path = "C:\\projects\\bedside-vlc\\VlcForm\\bin\\Debug\\netcoreapp3.1\\VlcForm.exe"
```
Comment (lmancel) [inline src/frontend/src/players/vlc/player.ts]: 🙅

File: src/launcher/index.ts [hunk 4]
```diff
@@ -123,6 +130,12 @@ async function main() {
         if (flowrWindow.store.get('clearAppDataOnStart')) {
           await clearBrowsingData()
         }
+
+        // If the browserWindow is not properly configured to use a player configured zPosition, destroy and remake it
+        if (flowrWindow.transparent && userPreferencesMerged.player.position !== PlayerPosition.BACKGROUND) {
+          app.quit();
+          app.relaunch();
```
Comment (lmancel) [inline src/launcher/index.ts]:
```suggestion
          app.relaunch();
          app.quit();
```

File: src/launcher/index.ts [hunk 5]
```diff
@@ -123,6 +130,12 @@ async function main() {
         if (flowrWindow.store.get('clearAppDataOnStart')) {
           await clearBrowsingData()
         }
+
+        // If the browserWindow is not properly configured to use a player configured zPosition, destroy and remake it
+        if (flowrWindow.transparent && userPreferencesMerged.player.position !== PlayerPosition.BACKGROUND) {
```
Comment (lmancel) [inline src/launcher/index.ts]: Shouldn't this also cover the opposite case ?
```suggestion
        if ((flowrWindow.transparent && userPreferencesMerged.player.position !== PlayerPosition.BACKGROUND)
            || !flowrWindow.transparent && userPreferencesMerged.player.position !== PlayerPosition.FOREGROUND
        ) {
```

**Review by lmancel — CHANGES_REQUESTED** [review]

File: src/launcher/index.ts [hunk 6]
```diff
@@ -19,7 +19,7 @@ import { keyboard } from '../keyboard/keyboardController'
 import { cloneDeep, mergeWith } from 'lodash'
 import { FullScreenManager } from '../common/fullscreen'
 import { IFlowrDesktopConfig } from '../frontend/src/interfaces/IFlowrDesktopConfig'
-import { PlayerPosition } from '../frontend/src/interfaces/playerStore'
+import { WexondOptions } from "../wexond/main/app-window";
```
Comment (lmancel) [inline src/launcher/index.ts]: We should add prettier to the commit hook as well
```suggestion
import { WexondOptions } from '../wexond/main/app-window'
```

**Review by lmancel — APPROVED** [review]
Nothing else but a tiny style issue. Can be merged without it though

**Review by Azevedo-Dani — APPROVED** [review]

**Pull request merged**